torch_xla/experimental/pallas_kernels/multi_queries_paged_attention_kernel.py

```diff
@@ -0,0 +1,612 @@
+  q = jnp.permute_dims(q, (0, 2, 1, 3))
+  num_kv_heads, _, page_size, head_dim_k = k_pages.shape
+  batch_size_paged_indices, pages_per_sequence = page_indices.shape
+
+  if k_pages.shape != v_pages.shape:
+    raise ValueError(
+        f"k_pages and v_pages must have the same shape. Got {k_pages.shape} and"
+        f" {v_pages.shape}"  # pytype: disable=attribute-error
+    )
+  if head_dim_k != head_dim:
+    raise ValueError("head_dim of Q must be the same as that of K/V. Got"
+                     f" {head_dim} and {head_dim_k}.")
+  if lengths.shape != (batch_size,):
+    raise ValueError("`lengths` and `q` must have the same batch size")
+  if batch_size_paged_indices != batch_size:
+    raise ValueError("`page_indices` and `q` must have the same batch size")
+  if lengths.dtype != jnp.int32:
+    raise ValueError(
+        f"The dtype of `lengths` must be int32. Got {lengths.dtype}")
+  if num_queries_per_compute_block > query_len:
+    raise ValueError(
+        f"{num_queries_per_compute_block=} should be smaller or equal to {query_len=}"
+    )
+  if num_kv_pages_per_compute_block > pages_per_sequence:
+    raise ValueError(
+        f"{num_kv_pages_per_compute_block=} should be smaller or equal to {pages_per_sequence=}"
+    )
+  if pages_per_sequence % num_kv_pages_per_compute_block != 0:
+    raise ValueError(
+        "num_kv_pages_per_compute_block must be divisible by pages per sequence. Got"
+        f" {pages_per_sequence=} and {num_kv_pages_per_compute_block=}.")
+  if num_q_heads % num_kv_heads != 0:
+    raise ValueError(
+        "Number of Q heads must be divisible by number of KV heads. Got"
+        f" {num_q_heads} and {num_kv_heads}.")
+  num_q_heads_per_kv_head = num_q_heads // num_kv_heads
+
+  # grid
+  grid = (
+      batch_size,
+      num_kv_heads,
+      # what if query_len%num_queries_per_compute_block!=0?
+      query_len //
+      num_queries_per_compute_block,  # how many compute blocks we need to loop the query_len
+      pages_per_sequence //
+      num_kv_pages_per_compute_block,  # how many compute blocks we need to loop the kv_len
+  )
+
+  # out_shape
+  o_shape = jax.ShapeDtypeStruct(shape=q.shape, dtype=q.dtype)
+  l = jax.ShapeDtypeStruct((batch_size, num_q_heads, query_len, MIN_BLOCK_SIZE),
+                           dtype=jnp.float32)
+  m = jax.ShapeDtypeStruct((batch_size, num_q_heads, query_len, MIN_BLOCK_SIZE),
+                           dtype=jnp.float32)
+  out_shape = (o_shape, l, m)
+
+  # in-spec. Note q.shape=[batch_size, num_q_heads, query_len, head_dim]
+  # Map from grid idx.
+  print(f'xw32 line591 {num_q_heads_per_kv_head=}')
+
+  def qo_index_map(batch_index, kv_head_index, q_seq_blk_idx, *_):
+    return (batch_index, kv_head_index, q_seq_blk_idx, 0)
+
+  q_block_spec = pl.BlockSpec(
+      # q_ref.shape=[1, num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim]
+      (1, num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim),
```

Comment:
Can this be `(None, num_q_heads_per_kv_head, num_queries_per_compute_block, head_dim)` to drop the first dim of size 1 inside the kernel?

Reply:
Yeah, I think it can. I'm thinking if we should possibly split on the batch dimension later, similar to what the flash attention pallas kernel does today. If so, we'd need to add this batch dim back.